[RFC] ASoC: SDCA: add support for HIDE entity properties

shumingfan opened the pull request:
This is a draft patch for SDCA HIDE entity. It tries to use the standard HID device to parse the raw data.
The codec or bus driver could receive the raw data from UMP protocol, and then call 'hid_input_report' to report events.

charleskeepax committed:
HID: core: Add bus define for SoundWire bus

SDCA (SoundWire Device Class for Audio) uses HID to convey
input events from peripheral devices. Add a bus define for the
SoundWire bus to prepare support for this.

Signed-off-by: Charles Keepax <ckeepax@opensource.cirrus.com>

lgirdwood commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+						"mipi-sdca-HIDTx-supported-report-ids",
+						hide->hidtx_ids,
+						hide->num_hidtx_ids);
+		if (ret < 0)
```
do we need to fwnode_handle_put() here ?

charleskeepax replied:
I don't think this is missing one I think the one above is incorrect, entity_node is requested and put in find_sdca_entities one doesn't need to manage its lifetime here.

shumingfan replied:
The find_sdca_entities function will call fwnode_handle_put, not here.
I will remove fwnode_handle_put call in this function.

lgirdwood commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+						"mipi-sdca-HIDRx-supported-report-ids",
+						hide->hidtx_ids,
+						hide->num_hidtx_ids);
+		if (ret < 0)
```
ditto re handle put.

lgirdwood commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,124 @@ static int find_sdca_entity_ge(struct device *dev,
 
+static int sdwhid_parse(struct hid_device *hid)
+{
+	//struct sdw_slave *slave = dev_to_sdw_dev(hid->dev.parent);
```
can be deleted.

shumingfan replied:
will delete if not needed

lgirdwood commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,124 @@ static int find_sdca_entity_ge(struct device *dev,
+	rsize = entity->hide.hid_desc.desc[0].wDescriptorLength;
+
+	if (!rsize || rsize > HID_MAX_DESCRIPTOR_SIZE) {
+		dev_err(&hid->dev, "weird size of report descriptor (%u)\n", rsize);
```
invalid size of ....

shumingfan replied:
will fix

lgirdwood commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,124 @@ static int find_sdca_entity_ge(struct device *dev,
+	return ret;
+}
+
+static int sdwhid_start(struct hid_device *hid)
+{
+	return 0;
+}
+
+static void sdwhid_stop(struct hid_device *hid)
+{
+
+}
```
are these still todo ?

shumingfan replied:
Yes, the HID driver will call hid_hw_start in the hid_device_probe.
The hid_hw_start function will call .start callback directly. If there is no this callback, the kernel will show NULL pointer error messages.
https://github.com/thesofproject/linux/blob/3fc50b3e6d9ff33e0a3c4079f53d2db9ee37ac72/drivers/hid/hid-core.c#L2349
However, I did the HID button test, and it worked with the empty .start/.stop callback.
I will check if something we can add to the .start/.stop callback function.

ujfalusi commented on include/sound/sdca_function.h:
```diff
@@ -971,6 +972,32 @@ struct sdca_entity_ge {
 
+/**
+ * struct sdca_entity_hide - information specific to HIDE Entities
+ * @*hid: HID device structure
```
the `*` is not needed in struct documentation.

Also, the hid is not used in this patch?

shumingfan replied:
The hid will be used in the codec driver or bus driver that gets the raw data from UMP protocol.
That driver will call hid_input_report to send the raw data.

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
 }
 
+static int find_sdca_entity_hide(struct device *dev, struct fwnode_handle *function_node,
+	struct fwnode_handle *entity_node, struct sdca_entity *entity)
```
nitpick: alignment:
```c
static int
find_sdca_entity_hide(struct device *dev, struct fwnode_handle *function_node,
		      struct fwnode_handle *entity_node,
		      struct sdca_entity *entity)
```

shumingfan replied:
will fix

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+	int nval, ret;
+	int num_hid_desc, num_report_decs;
+	struct hid_descriptor hid_desc;
+	unsigned char *report_desc = NULL;
```
nitpick: can you resort these and combine the same types

shumingfan replied:
will fix

plbossart replied:
and one line per variable is you have a mix of array, pointers and scalars.

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+	if (ret)
+		hide->max_delay = 0;
+	else
+		hide->max_delay = delay;
```
probably it is cleaner:
```c
hide->max_delay = ret ? 0 : delay;
```
?

charleskeepax replied:
max_delay will already be zero initialised when sdca_entity is allocated so just:
if (!ret)
  hide->max_delay = delay;
Would get my vote, Mark also tends not to be the biggest fan of ternaries.

shumingfan replied:
will fix

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+
+	/* register HID device */
+	num_hid_desc = fwnode_property_count_u8(function_node, "mipi-sdca-hid-descriptor");
+	fwnode_property_read_u8_array(function_node, "mipi-sdca-hid-descriptor", (u8 *)&hid_desc, num_hid_desc);
```
is it really exactly `struct hid_descriptor` which is adopted by MIPI and is in the ACPI table???

charleskeepax replied:
As far as I understand things yes that is correct.

shumingfan replied:
Yes, it is the same structure as a standard HID descriptor.

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+	fwnode_property_read_u8_array(function_node, "mipi-sdca-hid-descriptor", (u8 *)&hid_desc, num_hid_desc);
+	hide->hid_desc = hid_desc;
+
+	num_report_decs = fwnode_property_count_u8(function_node, "mipi-sdca-report-descriptor");
```
Can you re-use the same variable for all these counts in the function?

shumingfan replied:
will fix

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+	hide->hid_desc = hid_desc;
+
+	num_report_decs = fwnode_property_count_u8(function_node, "mipi-sdca-report-descriptor");
+	report_desc = kzalloc(num_report_decs, GFP_KERNEL);
```
this is going to leak memory? devm_kzalloc() ?

ujfalusi replied:
can num_report_decs be 0 ?

shumingfan replied:
will fix

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+	/* register HID device */
+	num_hid_desc = fwnode_property_count_u8(function_node, "mipi-sdca-hid-descriptor");
+	fwnode_property_read_u8_array(function_node, "mipi-sdca-hid-descriptor", (u8 *)&hid_desc, num_hid_desc);
+	hide->hid_desc = hid_desc;
```
I would prefer memcpy() if possible, first I thought that you are going to save the pointer of a local struct before I went back and check the structs and types...

shumingfan replied:
will fix

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+	fwnode_property_read_u32_array(entity_node, "mipi-sdca-hide-related-audio-function-list",
+		&af_list[0], num_af);
+
+
```
extra line break.

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,92 @@ static int find_sdca_entity_ge(struct device *dev,
+		&af_list[0], num_af);
+
+
+	/* register HID device */
```
I'm not sure, but I don't see that a HID device is registered...

shumingfan replied:
I will change the API to 'add_sdw_hid_device'.

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,124 @@ static int find_sdca_entity_ge(struct device *dev,
+	kfree(rdesc);
+	if (ret) {
+		dev_err(&hid->dev, "parsing report descriptor failed\n");
+		goto err;
```
just return from here, or
```c
if (!ret)
	return 0;

dev_err(&hid->dev, "parsing report descriptor failed\n");
return ret;
```

shumingfan replied:
will fix

ujfalusi commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,124 @@ static int find_sdca_entity_ge(struct device *dev,
+
+static int sdwhid_start(struct hid_device *hid)
+{
+	return 0;
```
these are mandatory and we don't need to do anything?

shumingfan replied:
Please check the comment above.

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1222,6 +1222,213 @@ static int find_sdca_entity_ge(struct device *dev,
+	entity->hide.hid = hid;
+
+	return 0;
+}
```
These actual HID helpers should probably not go in this file, this file is all about parsing the DisCo stuff. I would be tempted to make this part of an SDCA HID module.

shumingfan replied:
OK, will move it to the SDCA HID module.

charleskeepax commented on sound/soc/sdca/sdca_hid.c:
```diff
@@ -0,0 +1,139 @@
+		return PTR_ERR(hid);
+
+	hid->ll_driver = &sdw_hid_driver;
+
```
We should probably set hid->bus here, I have done a patch that adds a SoundWire define, please feel free to pull it into this series: https://github.com/charleskeepax/linux/commit/0cc291efe6e0e6d65d178bd6e2f5dd38993cd680

charleskeepax commented on sound/soc/sdca/sdca_hid.c:
```diff
@@ -0,0 +1,139 @@
+	hid->dev.parent = dev;
+
+	hid->name[0] = 0;
+	if (!strlen(hid->name))
```
This is a bit odd after setting name[0] = 0, then strlen will always equal 0.

charleskeepax commented on sound/soc/sdca/sdca_hid.c:
```diff
@@ -0,0 +1,139 @@
+{
+	struct sdw_bus *bus;
+	struct hid_device *hid;
+	struct sdw_slave *slave = dev_to_sdw_dev(dev);
```
Hmm... although actually maybe this is fine, probably ignore this comment for now it kinda depends on how you create the class driver at the end.

charleskeepax commented:
Ok have run this up on our devices and it seems to work well, I think the overall approach looks pretty good to me. Thank you very much for contributing these changes. I will go through today and do a little more review of the minor details but I think from a high level this looks good to me.

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1220,6 +1221,98 @@ static int find_sdca_entity_ge(struct device *dev,
+			return ret;
+	} else {
+		hide->hidtx_ids = NULL;
+	}
```
Don't need to clear this here, it will already be zeroed.

shumingfan replied:
will fix

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1220,6 +1221,98 @@ static int find_sdca_entity_ge(struct device *dev,
+	nval = fwnode_property_count_u32(entity_node, "mipi-sdca-HIDRx-supported-report-ids");
+	if (nval > 0) {
+		hide->num_hidtx_ids = nval;
+		hide->hidtx_ids = devm_kcalloc(dev, hide->num_hidtx_ids,
```
These should be num_hidrx_ids, not tx.

shumingfan replied:
will fix

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1220,6 +1221,98 @@ static int find_sdca_entity_ge(struct device *dev,
+		if (ret < 0)
+			return ret;
+	} else {
+		hide->hidrx_ids = NULL;
```
Again no need to clear here.

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1220,6 +1221,98 @@ static int find_sdca_entity_ge(struct device *dev,
+
+	nval = fwnode_property_count_u32(entity_node, "mipi-sdca-hide-related-audio-function-list");
+	if (nval <= 0) {
+		dev_err(dev, "%pfwP: HIDE - audio function numbers list missing: %d\n",
```
No need to put HIDE in the message here the entity part will make it clear enough this is hid.

shumingfan replied:
will fix

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1220,6 +1221,98 @@ static int find_sdca_entity_ge(struct device *dev,
+	}
+
+	fwnode_property_read_u32_array(entity_node, "mipi-sdca-hide-related-audio-function-list",
+		&af_list[0], nval);
```
Just af_list, it should be equivalent to &af_list[0].

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1220,6 +1221,98 @@ static int find_sdca_entity_ge(struct device *dev,
+	nval = fwnode_property_count_u8(function_node, "mipi-sdca-hid-descriptor");
+	if (nval)
+		fwnode_property_read_u8_array(function_node, "mipi-sdca-hid-descriptor", (u8 *)&hid_desc, nval);
+	memcpy(&hide->hid_desc, &hid_desc, sizeof(hid_desc));
```
Not sure this memcpy adds anything why not just read into the hide->hid_desc buffer directly?

charleskeepax replied:
Also I think the hid-descriptor property is required for HID functions so we should probably error if it isn't here.

shumingfan replied:
Thanks for reviewing. Will fix.

charleskeepax commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1220,6 +1221,98 @@ static int find_sdca_entity_ge(struct device *dev,
+
+			/* add HID device */
+			sdca_add_hid_device(dev, entity_node, entity);
+		}
```
Ditto here, pretty sure report-descriptor is required for HIDs so we should error out.

shumingfan replied:
The patch calls 'fwnode_property_count_u8' to check whether the report descriptor exists.
So the code will add the HID device when the HID/Report descriptors exist.

plbossart replied:
I am with @charleskeepax, we need to make sure all reads are tested. Checking the size of the array is not good enough. we had a similar discussion for mipi properties with @andy-shev a long time ago. Don't assume all reads succeed.

charleskeepax replied:
For clarity, I think the check on the call to count should be enough to ensure the read itself succeeds, unless I have missed something. My comment was really more that if the entity is a HIDE one (which is is if we are here in the code) then "mipi-sdca-report-descriptor" isn't optional, so we should return an error if it doesn't exist.

charleskeepax commented on sound/soc/sdca/sdca_hid.c:
```diff
@@ -0,0 +1,133 @@
+	if (!rdesc)
+		return -ENOMEM;
+
+	memcpy(rdesc, entity->hide.hid_report_desc, rsize);
```
Is there a problem with just passing the hid_desc in directly, doesn't look like we need to copy it to me?

shumingfan replied:
ok, sure. will fix.

charleskeepax commented on sound/soc/sdca/sdca_hid.c:
```diff
@@ -0,0 +1,133 @@
+
+static void sdwhid_close(struct hid_device *hid)
+{
+
```
Remove the blank line in this stub.

shumingfan replied:
will fix

charleskeepax commented on sound/soc/sdca/sdca_hid.c:
```diff
@@ -0,0 +1,133 @@
+
+	ret = hid_add_device(hid);
+	if (ret) {
+		if (ret != -ENODEV)
```
Perhaps combine these if (ret && ret != -ENODEV)

charleskeepax commented on sound/soc/sdca/sdca_hid.c:
```diff
@@ -0,0 +1,133 @@
+};
+
+int sdca_add_hid_device(struct device *dev, struct fwnode_handle *entity_node,
+	struct sdca_entity *entity)
```
No need to pass in entity_node here it is unused.

shumingfan replied:
will fix

charleskeepax reviewed (approved):
Could use a few tabbing/alignment fixes, but basically looks good to me now. Thanks again for working on this.

shumingfan commented:
> Could use a few tabbing/alignment fixes, but basically looks good to me now. Thanks again for working on this.

will fix

plbossart commented on sound/soc/sdca/sdca_functions.c:
```diff
@@ -1220,6 +1220,85 @@ static int find_sdca_entity_ge(struct device *dev,
+	}
+
+	nval = fwnode_property_count_u32(entity_node, "mipi-sdca-hide-related-audio-function-list");
+	if (nval <= 0) {
```
Are you sure Function0 is an error? I mean, we have a 3-bit code for functions et we can have 8 functions, so zero must be included, no?

Another point is that you'd want to verify is each element in this list points to a real function, so avoid problems with inconsistent APCI stuff.

shumingfan replied:
I think the function number should be added to this list, even Function0. 
This function parses the properties and collects the data. It will be verified when it is used.

plbossart commented on sound/soc/sdca/sdca_hid.c:
```diff
@@ -0,0 +1,124 @@
+static int sdwhid_parse(struct hid_device *hid)
+{
+	struct sdca_entity *entity = hid->driver_data;
+	unsigned int rsize = 0;
```
useless init, see line 27

plbossart commented on sound/soc/sdca/sdca_hid.c:
```diff
@@ -0,0 +1,124 @@
+{
+	switch (reqtype) {
+	case HID_REQ_GET_REPORT:
+		/* not implemented yet */
```
and IMHO this implementation should rely on a common library to deal with UMP TX and RX. I vaguely remember suggesting something before I left Intel, not sure if it was kept somewhere. @bardliao any memories of this?

charleskeepax replied:
Seems reasonable to have UMP library helpers but this code only registers the HID stuff so it doesn't actually interface with any of the UMP bits yet, so I don't think that needs to be implemented for this series. We really need to merge the IRQ stuff before we get to the UMP interactions (series should be going out on Monday assuming the merge window closes on time).

bardliao replied:
> and IMHO this implementation should rely on a common library to deal with UMP TX and RX. I vaguely remember suggesting something before I left Intel, not sure if it was kept somewhere. @bardliao any memories of this?

@plbossart I can find https://github.com/thesofproject/linux/pull/5032. Is it what you are referring to?

shumingfan committed:
ASoC: SDCA: add support for HIDE entity properties and HID descriptor/report

Add support for parsing the HIDE entity descriptor and HID descriptor/report

Signed-off-by: Shuming Fan <shumingf@realtek.com>

shumingfan committed:
ASoC: SDCA: add a HID device for HIDE entity

This patch supports to add a HID device for SDCA HIDE entity.
The codec driver could call 'hid_input_report' to report events.

Signed-off-by: Shuming Fan <shumingf@realtek.com>

charleskeepax reviewed (approved):
Personally, I am good for you to send a version of this series upstream be good to see if there are any comments from the wider community on this.

shumingfan commented:
> Personally, I am good for you to send a version of this series upstream be good to see if there are any comments from the wider community on this.

OK, sure. But, we need to send your patch (https://github.com/thesofproject/linux/pull/5429/commits/dd7969e5bbda8d6f75223ca4c6b25cc4cbc0916a) first.
Would you send that patch, or should I send it?

charleskeepax commented:
I am happy for you to send it, makes sense for it all to be in the same series, as long as you are also ok with that?

shumingfan commented:
> you send that p

OK, I will send these patches in the same series.
